### PR TITLE
Add data pod plugin infrastructure

### PR DESCRIPTION
**Pull request by rudigiesler**
This ticket is for adding infrastructure for installing data pod types as plugins and looking up the available pod types. Supporting creation and management of instances of these pod types is out of this ticket's scope.

### CODE REVIEW

**Comment by rudigiesler** [comment]
@justinvdm Ready for a first look.

File: casepro/pods/registry.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    '''
+    Given the index of the pod, and the config dictionary for the pod, this
+    returns the instance of that pod.
+    '''
+    class_name = config.pop('type')
+    config['index'] = index
+    return get_class_from_string(class_name)(config)
+
+
+pods = [
```
Comment (hodgestar) [inline casepro/pods/registry.py]: Maybe this should be a tuple rather than a list? Appending to pods later will result in URL patterns not being set up correctly.

Comment (hodgestar) [inline casepro/pods/registry.py]: I'm wondering whether this should be a function rather than a global variable in case we want to make it more dynamic later. Not sure how to make that work with URL patterns though.

File: casepro/pods/registry.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+
+def get_url_patterns():
+    '''
+    Get the list of URL patterns for all of the pods.
+    '''
+    return [
+        pattern
+        for pod in pods
+        for pattern in pod.url_patterns
```
Comment (hodgestar) [inline casepro/pods/registry.py]: If we have multiple pods of the same class, this will result in duplicate URL patterns. Maybe we should only load URL patterns per-class or something?

File: casepro/pods/base.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        "Automatically determined and set in the pod registry.",
+        required=True)
+
+
+class Pod(object):
+    '''
+    The base class for all pod plugins.
+    '''
+    config_cls = PodConfig
+    url_patterns = []
```
Comment (hodgestar) [inline casepro/pods/base.py]: Maybe this should be a tuple too?

File: casepro/pods/registry.py (hunk 2 above)
Comment (rudigiesler) [inline casepro/pods/registry.py]: My idea here was that since the pods have their `index`, they could use that to namespace their own URLs. Or maybe we should namespace it for them? But then you run into the issue of multiple URLs having the same `name`.

File: casepro/pods/registry.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+
+def load_pod(index, config):
+    '''
+    Given the index of the pod, and the config dictionary for the pod, this
+    returns the instance of that pod.
+    '''
+    class_name = config.pop('type')
+    config['index'] = index
+    return get_class_from_string(class_name)(config)
```
Comment (justinvdm) [inline casepro/pods/registry.py]: It would be nice to be able to leverage `django.apps.get_app_configs()` to load the pods like we did in the test branch, would avoid the need to be loading classes from strings ourselves and would mean we can use `INSTALLED_APPS` to register pod types. That will allow us to include each pod's static files in the directories django looks for static files in without needing to add any boilerplate.

File: casepro/pods/registry.py (hunk 2 above)
Comment (justinvdm) [inline casepro/pods/registry.py]: From the conversation we had at the end of Monday, I think both you and hodgestar suggested we register a single url route for all pods and inside that route just delegate to a method on the pod class. Wondering whether we should go that route instead?

File: casepro/pods/base.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    name = 'Default pod name'
+    controller = None
+    directive = None
```
Comment (justinvdm) [inline casepro/pods/base.py]: We'll need to add configuration for paths to scripts and stylesheets here, but probably makes more sense to do that in a separate PR (happy to take that on once this PR lands).

File: pip-freeze.txt [hunk 6]
```diff
@@ -8,6 +8,7 @@ billiard==3.3.0.22
 boto==2.38.0
 celery==3.1.19
 colorama==0.3.7
+confmodel==0.2.0
```
Comment (justinvdm) [inline pip-freeze.txt]: I'm guessing this is already the plan, but just in case, we should ask the relevant people whether they are alright with us including this dependency.

**Comment by justinvdm** [comment]
Left a few comments, looks good so far.

File: casepro/pods/registry.py (hunk 1 above)
Comment (rudigiesler) [inline casepro/pods/registry.py]: Changed to tuple in https://github.com/praekelt/casepro/pull/31/commits/10f4b67e9d5d618f8c256c7b850bdc54a489ddab

File: casepro/pods/base.py (hunk 3 above)
Comment (rudigiesler) [inline casepro/pods/base.py]: Addressed in https://github.com/praekelt/casepro/pull/31/commits/10f4b67e9d5d618f8c256c7b850bdc54a489ddab

File: casepro/pods/registry.py (hunk 4 above)
Comment (rudigiesler) [inline casepro/pods/registry.py]: I've changed the way that this works here: https://github.com/praekelt/casepro/pull/31/commits/cb691feb4a33172c625fd4bca1a00e483e7755be

File: casepro/pods/base.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    def url_patterns(self):
+        return (
+            url(r'read/%d/$' % self.config.index, self.request_callback),
+        )
+
+    def request_callback(self, request):
+        '''
+        Called to get the details that should appear within the pod.
+        '''
+        return HttpResponse()
```
Comment (hodgestar) [inline casepro/pods/base.py]: Should this perhaps raise `NotImplementError` or return JSON?

File: casepro/pods/registry.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    return appconfig.pod_class
+
+
+def load_pod(index, config):
+    '''
+    Given the index of the pod, and the config dictionary for the pod, this
+    returns the instance of that pod.
+    '''
+    config = config.copy()
+    app_label = config.pop('type')
```
Comment (hodgestar) [inline casepro/pods/registry.py]: Maybe we should name this key `app_label` so its clear what it is?

File: casepro/pods/base.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    controller = None
+    directive = None
+
+    def __init__(self, config):
+        self.config = self.config_cls(config)
+
+    @property
+    def url_patterns(self):
+        return (
+            url(r'read/%d/$' % self.config.index, self.request_callback),
```
Comment (hodgestar) [inline casepro/pods/base.py]: There needs to be a way for the client side to figure out what this URL is.

@justinvdm mentioned the idea of having two fixed URLs (that we all discussed previously) that then just map to the appropriate data pod and call a method there. I'm still leaning towards that as a better idea that allowing complex arbitrary URLs (although maybe for very complex pods the ability to add custom URLs per pod-type would still be useful). I'd be happy to leave that our entirely until we encounter the need though.

Comment (rudigiesler) [inline casepro/pods/base.py]: Would the data pod do something like add its index to the request, which we can then route on? If the data pod knows its index, then is there much difference between it putting it in the URL or putting it in the query args? I was looking at this for contructing the urls (although maybe I'm understanding it incorrectly): https://github.com/praekelt/casepro/pull/34/files#diff-605d5fe78f6a6aac57fb0e52c1ecceb3R428

File: casepro/pods/base.py (hunk 7 above)
Comment (rudigiesler) [inline casepro/pods/base.py]: Addressed in https://github.com/praekelt/casepro/pull/31/commits/37d76de2a1ec4487f9649a61a0bb63fc4daddf0a

File: casepro/pods/registry.py (hunk 8 above)
Comment (rudigiesler) [inline casepro/pods/registry.py]: Addressed in https://github.com/praekelt/casepro/pull/31/commits/bcc8013c2587fb612ee6a24b12c8572b95c2fa76

File: casepro/pods/base.py (hunk 9 above)
Comment (hodgestar) [inline casepro/pods/base.py]: I thinking of having a `urls.py` that looks like:
```python
url_patterns = [
   url(r'read/%(pk)s/$', views.read_pod_data),
   url(r'action/(%pk)s/$`, views.perform_pod_action),
]
```
and then a `views.py` that looks like:
```python
def read_pod_data(request, pk):
    pod = get_pod(pk)
    data = pod.read_data()
    return HttpResponse(json.dumps(data))

def perform_pod_action(request, pk):
    pod = get_pod(pk)
    params = json.loads(request.body)
    result = pod.perform_action(params)
    return HttpResponse(json.dumps(data))
```

Comment (rudigiesler) [inline casepro/pods/base.py]: Ah I see, yes we can do it like that.

Comment (rudigiesler) [inline casepro/pods/base.py]: Done in https://github.com/praekelt/casepro/pull/31/commits/4b6878af31514be16d992fcf64d98f5939e73525

**Comment by rudigiesler** [comment]
@justinvdm @hodgestar Ready for another review.

**Comment by justinvdm** [comment]
:+1: from me when travis is happy, looks good :)

Not sure if we want to wait to hear if people are okay with us adding `confmodel` as dependency?

File: casepro/pods/registry.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+
+def get_url_patterns():
+    '''
+    Get the list of URL patterns for all of the pods.
+    '''
+    return tuple(
+        pattern
+        for pod in pods
+        for pattern in pod.url_patterns
```
Comment (hodgestar) [inline casepro/pods/registry.py]: I still think it is more likely that custom routes will be per-pod-type, rather than per pod. Just flagging this though -- not necessarily important to address in this PR.

Comment (rudigiesler) [inline casepro/pods/registry.py]: Hmm.. That means that the url patterns should go in the app config (PodPlugin) rather than on the class?

How do django apps usually handle this? Do they have a `urls.py` that they use to add the urls in `settings.py`? If that is the case, maybe we should use that mechanism, and just remove all the custom URL things in this PR?

**Comment by rudigiesler** [comment]
@justinvdm @hodgestar I've added an extra test for the invalid JSON case.

Also, thoughts in https://github.com/praekelt/casepro/pull/31#discussion_r70646379 ? Should we remove the custom URL things all together? The issue is that it's not something that can be added to `settings.py` (as far as I can figure out), you have to add the urls to `urls.py`.

**Comment by justinvdm** [comment]
+1 on removing custom urls, we can add it later if we need it.

**Comment by rudigiesler** [comment]
@justinvdm @hodgestar I've remove the custom URL things in https://github.com/praekelt/casepro/pull/31/commits/51d1b90bc6f878b8c386a2000beead7f73737b54 . Ready for a (hopefully) final review.

**Comment by hodgestar** [comment]
:+1: from me.

**Comment by justinvdm** [comment]
:+1: from me
